server/src/routes/instances.js

```diff
@@ -0,0 +1,148 @@
+import express from 'express'
+import * as R from 'ramda'
+import {
+  // getRunningContainers,
+  getRunningContainersByUser,
+  createNewContainer,
+  stopContainer,
+  deleteContainer,
+  getContainerLogs,
+} from '../docker'
+import { verifyUser, resolveContainer } from '../database'
+import { attempt, uuid } from './helpers'
+
+const router = express.Router()
+
+
+// @api GET /api/instances
+// Retrieve all instances in the database.
+//
+// Returns 200 OK and a JSON array of all instances or 500 ISE if an error occurred.
+/*
+router.get('/', async (req, res) => {
+  try {
+    const containers = await getRunningContainers()
+    res.send(containers)
+  } catch (error) {
+    console.log(error)
+    res.sendStatus(500)
+  }
+})
+*/
+router.get('/', async (req, res) => {
+  try {
+    const [type, token] = req.get('Authorization').split(' ')
+    // let users = JSON.parse(atob(token))
+    if (type !== 'Feedbacker') throw new Error('Unsupported authorization type')
+    let users = JSON.parse(Buffer.from(token, 'base64').toString())
+    users = R.toPairs(users)
+    const containers = []
+    for (const [userId, secret] of users) {
+      try {
+        await verifyUser(userId, secret)
+        containers.push(...await getRunningContainersByUser([userId]))
+      } catch (error) { /* ignore */ }
+    }
+    res.send(containers)
+  } catch (error) {
+    console.log(error)
+    res.sendStatus(500)
+  }
+})
+
+// @api GET /api/instances/logs/:name
+// Retrieve logs of an instance.
+//
+// Returns 200 OK and a string with logs or 500 ISE if an error occurred.
+router.get('/logs/:name', async (req, res) => {
+  try {
+    const id = await resolveContainer(req.params.name)
+
+    const logs = await getContainerLogs(id)
+    res.type('txt')
+    res.send(logs)
+  } catch (error) {
+    console.log(error)
+    res.sendStatus(500)
+  }
+})
+
+// @api POST /api/instances/new
+// Create a new container.
+//
+// Currently the only parameter considered is `instance_image`. The name and subdomain are
+// generated automatically.
+//
+// Example body @json {
+//  "type": "node"
+// }
+//
+// Returns 200 OK if the operation completed successfully and 500 ISE if an error occurred.
+router.post('/new', async (req, res) => {
+  try {
+    const {
+      url, version, type, name, port, userId,
+    } = req.body
+
+    if (name.length < 3 || name.length > 20) {
+      throw new Error(`Name too short or long ${name}`)
+    }
+    if (!name.match(/[a-z0-9](-?[a-z0-9])*/)) {
+      throw new Error(`Bad container name ${name}`)
```

Comment:
technically these should result in status code 400 instead of 500

Reply:
Fixing it ATM, I made a `HttpError` type with a status prop

Reply:
Preview:
```js
  if (name.length < 3 || name.length > 20) {
    throw new HttpError(400, `Name too short or long ${name}`)
  }
  if (!name.match(/[a-z0-9](-?[a-z0-9])*/)) {
    throw new HttpError(400, `Bad container name ${name}`)
  }
```